cmd/snap/cmd_run.go

```diff
@@ -482,29 +473,103 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
+func maybeWaitWhileInhibited(snapName string, appName string) (info *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock, err error) {
+	if features.RefreshAppAwareness.IsEnabled() {
+		flow := newInhibitionFlow(snapName)
+		notified := false
+		notInhibited := func() (err error) {
+			// get updated "current" snap info
+			info, app, err = getInfoAndApp(snapName, appName, snap.R(0))
+			return err
+		}
+		inhibited := func(hint runinhibit.Hint, inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
+			if !notified {
+				info, app, err = getInfoAndApp(snapName, appName, inhibitInfo.Previous)
+				if err != nil {
+					return false, err
+				}
+				// don't wait, continue with old revision
+				if app.IsService() {
+					return true, nil
+				}
+				// don't start flow if we are not inhibited for refresh
+				if hint != runinhibit.HintInhibitedForRefresh {
+					return false, nil
+				}
+				// start notification flow
+				if err := flow.Notify(); err != nil {
+					return true, err
+				}
+				// make sure we call notification flow only once
+				notified = true
+			}
+			return false, nil
+		}
+
+		// If the snap is inhibited from being used then postpone running it until
+		// that condition passes. Inhibition UI can be dismissed by the user, in
+		// which case we don't run the application at all.
+		hintFlock, err = waitWhileInhibited(snapName, notInhibited, inhibited, 500*time.Millisecond)
+		if err != nil {
+			// it is fine to return an error here without finishing the notification
+			// flow because we either failed because of it or before it, so it
+			// should not have started in the first place.
+			return nil, nil, nil, err
+		}
+
+		if !app.IsService() && notified {
+			if err := flow.Finish(); err != nil {
```

Comment:
I see, this is why the `flow` needs to be defined outside of the `inhibited` callback. Thanks!